src/walker.ts

```diff
@@ -0,0 +1,99 @@
+import {
+  PotentialStep,
+  Pipeline,
+  Conditional,
+  Command,
+  CommandStep,
+  Step,
+} from '.';
+import { getAndCacheDependency } from './conditional';
+import { cloneDeep } from 'lodash';
+import { unwrapSteps } from './unwrapSteps';
+export interface Mutators {
+  pipelineFn?: (pipeline: Pipeline) => Promise<Pipeline>;
+  stepFn?: (step: Step) => Promise<Step>;
+  commandFn?: (command: Command) => Promise<Command>;
+}
+
+/**
+ * @param p The pipeline to evaluate
+ * @returns An evaluated pipeline with all conditionals generated if they were accepted
+ */
+export async function evaluatePipeline(p: Pipeline): Promise<Pipeline> {
+  const conditionalCache = new Map<any, any>();
+  p.steps = await unwrapSteps(p.steps, conditionalCache);
```

Comment:
I think you should use `sortedSteps` here, I looked at the implementation and Joscha was right: it already sorts the steps topologically, so if you iterate over them you'll visit them in order and you won't have to keep track of which ones you've visited and which ones you haven't.

Plus it prunes orphan steps that are effects of conditional steps that have not been accepted. It has quite a bit of logic for that that you wouldn't want to duplicate here.

Reply:
I just realised that if you add new edges between the existing nodes some of the effects might become required, so we need to do the pruning after the mutation anyway.

Reply:
I thought about it more: we don't want any changes to dependencies that will affect the iteration order anyway, because then we'll end up iterating on nodes before their dependencies.